stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+
+        Args:
+            dest (path) : The destination file/folder path within same project
+
+        Raises:
+            ValueError : When attempting to move projects
+            DNAnexusError : If attempting to move across projects
+        """
+        if not self.resource:
+            raise ValueError('Cannot move project ({})'.format(self))
+        if dest.canonical_project != self.canonical_project:
+            # This can be implemented by clone and remove original
+            raise DNAnexusError('Cannot move across different projects')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(new_dest))
+            folder_dest = dest
+        else:
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        file_handler.move(folder_dest.resource or '/')
+
+        if not dest_exists and not dest.endswith('/'):
+            file_handler.rename(dest.name)
+
+    def copy(self, dest, move_within_project=False):
+        """Copies data object to destination path.
+
+        Args:
+            dest (path|str): The destination file or directory.
+            move_within_project (bool): If True, make new files in DX instead of cloning.
+                Only comes into effect when both source and destination are DXPaths
+
+        Raises:
+            DNAnexusError : When cloning within same project with copy_new=False
+            DuplicateError : When cloning data object or folder to already existing path
+        """
+        dest = Path(dest)
+        if utils.is_dx_path(dest):
+            if self.isfile():
+                if dest.canonical_project == self.canonical_project:
+                    if move_within_project:
+                        self._move(dest)
+                    else:
+                        raise DNAnexusError('Cannot copy within same project without flag set')
+                else:
+                    self._clone(dest)
+            else:
+                raise NotFoundError('No data object was found for the given path on DNAnexus')
+        else:
+            super(DXPath, self).copy(dest)
 
     def download_object(self, dest):
-        """Download a single path or object to file."""
-        raise NotImplementedError
+        """Download a single path or object to file.
+
+        Args:
+            dest (Path): The output file
+
+        Raises:
+            ValueError: When source path is not a file
+        """
+        if not self.isfile():
```

Comment:
why is this check necessary? what is this causing?

Also wouldn't this fall under the same set of parameters as the other in terms of only checking that there is a resource, and then just treating it as a file until we find out it doesn't exist?

---

More broadly, on all of these messages - what are the standard things to raise?

* NotFoundError - if file or project doesn't exist
* UnauthorizedError - if credentials don't work (Have you been able to trigger this)

Reply:
also does this need a check that dest is a posixpath?

Reply:
You're absolutely right, I dont exactly remember why I added it there. And all my tests seem to pass without it, so I removed the check.
I haven't been able to trigger UnauthorizedError yet. Will get to that with error parsing.

And no I don't think, it needs to check that dest is a posixpath. Not only do the other obs modules not do that, but I think this might/will work on a windows path too.